ext/spl/spl_fixedarray.c

```diff
@@ -582,6 +587,67 @@ PHP_METHOD(SplFixedArray, __wakeup)
 	}
 }
 
+PHP_METHOD(SplFixedArray, __serialize)
+{
+	spl_fixedarray_object *intern = Z_SPLFIXEDARRAY_P(ZEND_THIS);
+	zval *current, tmp;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	array_init(return_value);
+
+	/* elements */
+	array_init_size(&tmp, intern->array.size);
+
+	for (zend_long i = 0; i < intern->array.size; i++) {
+		current = &intern->array.elements[i];
+		zend_hash_next_index_insert(Z_ARRVAL(tmp), current);
+		Z_TRY_ADDREF_P(current);
+	}
+
+	zend_hash_next_index_insert(Z_ARRVAL_P(return_value), &tmp);
+
+	/* members */
+	ZVAL_ARR(&tmp, zend_proptable_to_symtable(
+		zend_std_get_properties(&intern->std), /* always_duplicate */ 1));
+	zend_hash_next_index_insert(Z_ARRVAL_P(return_value), &tmp);
```

Comment:
I'd forgotten about the issues with serialize and SplFixedArray

For backwards compatibility, objects serialized in php <= 8.1 would need to be serialized in a way compatible with PHP >= 8.2, so that applications using SplFixedArray can be migrated without corrupting data (shared in memcached, redis, databases/files, etc)

- So instead of serializing `[list $fixedArrayFields, array $memberProperties]`, this should serialize `[...$fixedArrayFields, 'stringifiedMemberProp1' => $value]`, etc

`__unserialize` should then check if the property name is a string or an integer before deciding where to insert the property

Keep in mind that SplFixedArray can be subclassed, so there are legitimate uses for declared member properties

```
php > $x = new SplFixedArray(1);
php > $x[0] = 100;
php > $x->prop = 101
php > ;
php > $x->{3} = 102;
php > echo serialize($x);
O:13:"SplFixedArray":3:{s:4:"prop";i:101;s:1:"3";i:102;i:0;i:100;}
php > var_export(unserialize(serialize($x)));  // we should probably improve on this and set the member properties instead of changing the array count
SplFixedArray::__set_state(array(
   0 => 101,
   1 => 102,
   2 => 100,
))
php > var_dump(unserialize(serialize($x))->count());
int(3)
```

Reply:
I think the reason for that original behavior was because `__wakeup` would not give it a way to distinguish between `$x->{3} = 102;` and `$x[0] = 100` (those get coerced to string when adding to the object member properties hash table?), so it gave up and treated all member properties the same way, but with the array being passed in we don't lose that information